### PR TITLE
Added smoothstep built-in function

### PR DESCRIPTION
**Pull request by Chaosus**
Personally I misses hermite interpolation. So many algorithms are depends on it. It should be in the core, cuz its common in video game programming. Unity also have it in the builtin Math class https://docs.unity3d.com/ScriptReference/Mathf.SmoothStep.html 
In Godot, currently it can be only used  in shaders(smoothstep) or in Tween's. This PR exposes it to core - ~~I named it smooth_lerp for easy search it within other lerp-functions and(I suppose) its just another type of linear interpolation.~~

image

~~I will add C# version later, (if this will be accepted)~~

### CODE REVIEW

**Comment by bojidar-bg** [comment]
Lerp is a contraction for `linear_interpolation`. Maybe it should be called `smoothstep`/`smooth_step`?

**Comment by Chaosus** [comment]
Renamed to smooth_step

**Comment by Chaosus** [comment]
Added C# version too

**Comment by Calinou** [comment]
How does `lerp()` combined with `ease()` compare with this?

**Review by neikeq — APPROVED** [review]
C# change looks good to me but is wrongly formatted.

File: modules/mono/glue/Managed/Files/Mathf.cs [hunk 1]
```diff
@@ -261,6 +261,11 @@ public static real_t Sinh(real_t s)
             return (real_t)Math.Sinh(s);
         }
 
+        public static real_t SmoothStep(real_t from, real_t to, real_t weight) {
```
Comment (neikeq) [inline modules/mono/glue/Managed/Files/Mathf.cs]: The opening brace should go into a new line.

File: modules/mono/glue/Managed/Files/Mathf.cs [hunk 2]
```diff
@@ -261,6 +261,11 @@ public static real_t Sinh(real_t s)
...
+            real_t x = Clamp((weight - from) / (to - from), (real_t)0.0, (real_t)1.0);
+		    return x * x * (3 - 2 * x);
```
Comment (neikeq) [inline modules/mono/glue/Managed/Files/Mathf.cs]: This line is wrongly indented.

File: modules/mono/glue/Managed/Files/Mathf.cs (hunk 1 above)
Comment (Chaosus) [reply modules/mono/glue/Managed/Files/Mathf.cs]: Fixed

File: modules/mono/glue/Managed/Files/Mathf.cs (hunk 2 above)
Comment (Chaosus) [reply modules/mono/glue/Managed/Files/Mathf.cs]: Fixed now ?

**Comment by Chaosus** [comment]
@Calinou I think its different functions, even if you somehow force the results to be equal its not as easy as single call

File: modules/mono/glue/Managed/Files/Mathf.cs (hunk 2 above)
Comment (neikeq) [reply modules/mono/glue/Managed/Files/Mathf.cs]: Yes! :)

**Comment by bojidar-bg** [comment]
@Calinou Tested with KmPlot, the closest ease + lerp combination should be:
 ```gdscript
lerp(a, b, ease(t, -1.6521))
```

Here are some plots of both, along with their derivatives. Notice that ease's derivative has a discontinuity at x=0.5, unlike smoothstep's derivative.

f: smoothstep | g: ease | (f - g) * 100
--- | --- | ---
image | image | image

File: doc/classes/@GDScript.xml [hunk 3]
```diff
@@ -972,6 +972,19 @@
...
+			<return type="float">
+			</return>
+			<argument index="0" name="from" type="float">
+			</argument>
+			<argument index="1" name="to" type="float">
+			</argument>
+			<argument index="2" name="weight" type="float">
+			</argument>
+			<description>
+				Hermite interpolation between [code]from[/code] and [code]to[/code] using [code]weight[/code]. Simular to [code]lerp[/code] but interpolates faster at the beginning and slower at the end.
```
Comment (Calinou) [inline doc/classes/@GDScript.xml]:
```suggestion
				Hermite interpolation between [code]from[/code] and [code]to[/code] using [code]weight[/code]. Similar to [code]lerp[/code], but interpolates faster at the beginning and slower at the end.
```

**Review by Calinou — DISMISSED** [review]

File: doc/classes/@GDScript.xml (hunk 3 above)
Comment (Chaosus) [reply doc/classes/@GDScript.xml]: Done !

File: doc/classes/@GDScript.xml [hunk 4]
```diff
@@ -972,6 +972,19 @@
...
+			<return type="float">
+			</return>
+			<argument index="0" name="from" type="float">
+			</argument>
+			<argument index="1" name="to" type="float">
+			</argument>
+			<argument index="2" name="weight" type="float">
+			</argument>
+			<description>
+				Hermite interpolation between [code]from[/code] and [code]to[/code] using [code]weight[/code]. Similar to [code]lerp[/code], but interpolates faster at the beginning and slower at the end.
```
Comment (merumelu) [inline doc/classes/@GDScript.xml]:
```suggestion
				Hermite interpolation between [code]from[/code] and [code]to[/code] using [code]weight[/code]. Similar to [method lerp], but interpolates faster at the beginning and slower at the end.
```

Comment (Chaosus) [reply doc/classes/@GDScript.xml]: Done !

File: modules/visual_script/visual_script_builtin_funcs.h [hunk 5]
```diff
@@ -67,6 +67,7 @@ class VisualScriptBuiltinFunc : public VisualScriptNode {
 		MATH_LERP,
 		MATH_INVERSE_LERP,
 		MATH_RANGE_LERP,
+		MATH_SMOOTH_STEP,
```
Comment (vnen) [inline modules/visual_script/visual_script_builtin_funcs.h]: I believe this breaks compatibility if put here: it should be put at the end of the list. You can see in the doc changes that all constants got new values, so if someone is using a function from below, suddenly the functions will be changed and it'll break the code.

Comment (Chaosus) [reply modules/visual_script/visual_script_builtin_funcs.h]: I just remember the time I've implemented wrapi/wrapf function and it was ok to putted them in the middle of that list.. but ok I will place it on the bottom

Comment (Chaosus) [reply modules/visual_script/visual_script_builtin_funcs.h]: Done !

**Comment by Chaosus** [comment]
Added is_equal_approx check in order to fix division by zero (or NaN for floats)

File: doc/classes/@GDScript.xml (hunk 3 above)
Comment (Chaosus) [reply doc/classes/@GDScript.xml]: @Calinou Please dont forget to approve changes ^^

**Comment by reduz** [comment]
nice addition, I wonder whether it should be "smooth_step" or "smoothstep" like in glsl..

**Comment by Chaosus** [comment]
Changed to smoothstep (as also suggested by Calinou in IRC)

File: modules/mono/glue/Managed/Files/Mathf.cs [hunk 6]
```diff
@@ -261,6 +261,13 @@ public static real_t Sinh(real_t s)
             return (real_t)Math.Sinh(s);
         }
 
+        public static real_t SmoothStep(real_t from, real_t to, real_t weight)
+        {
+            if(IsEqualApprox(from, to)) return from;
```
Comment (akien-mga) [inline modules/mono/glue/Managed/Files/Mathf.cs]: Style issue here I guess, should have a space after `if`.

File: modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml [hunk 7]
```diff
@@ -204,7 +204,10 @@
 		<constant name="COLORN" value="62" enum="BuiltinFunc">
 			Return the [Color] with the given name and alpha ranging from 0 to 1. Note: names are defined in color_names.inc.
 		</constant>
-		<constant name="FUNC_MAX" value="63" enum="BuiltinFunc">
+		<constant name="MATH_SMOOTHSTEP" value="63" enum="BuiltinFunc">
+			Hermite interpolation.
```
Comment (akien-mga) [inline modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: Would be good to add more details about what this means.

File: modules/mono/glue/Managed/Files/Mathf.cs (hunk 6 above)
Comment (Chaosus) [reply modules/mono/glue/Managed/Files/Mathf.cs]: Fixed

File: modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml (hunk 7 above)
Comment (Chaosus) [reply modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: Fixed to "Return a number smoothly interpolated between the first two inputs, based on the third input. Using Hermite interpolation formula : float t; t = clamp((weight - from) / (to - from), 0.0, 1.0); return t * t * (3.0 - 2.0 * t);"

File: modules/visual_script/visual_script_builtin_funcs.h [hunk 8]
```diff
@@ -101,6 +101,7 @@ class VisualScriptBuiltinFunc : public VisualScriptNode {
 		VAR_TO_BYTES,
 		BYTES_TO_VAR,
 		COLORN,
+		MATH_SMOOTHSTEP,
```
Comment (akien-mga) [inline modules/visual_script/visual_script_builtin_funcs.h]: Is it a problem if it's included further up before `MATH_DECTIME`, as for `core/math`?

Comment (Chaosus) [reply modules/visual_script/visual_script_builtin_funcs.h]: by @vnen
 > I believe this breaks compatibility if put here: it should be put at the end of the list. You can see in the doc changes that all constants got new values, so if someone is using a function from below, suddenly the functions will be changed and it'll break the code.

File: modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml [hunk 9]
```diff
@@ -204,7 +204,10 @@
 		<constant name="COLORN" value="62" enum="BuiltinFunc">
 			Return the [Color] with the given name and alpha ranging from 0 to 1. Note: names are defined in color_names.inc.
 		</constant>
-		<constant name="FUNC_MAX" value="63" enum="BuiltinFunc">
+		<constant name="MATH_SMOOTHSTEP" value="63" enum="BuiltinFunc">
+			Return a number smoothly interpolated between the first two inputs, based on the third input. Using Hermite interpolation formula : float t; t = clamp((weight - from) / (to - from), 0.0, 1.0); return t * t * (3.0 - 2.0 * t);
```
Comment (akien-mga) [inline modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: - No space before `:`
- Simpler with `float t = clamp(...` directly
- For those not familiar with C++, GDScript syntax might be clearer
- Use `[code]` formatting, or better, a `[codeblock]`:
```
[codeblock]
var t = clamp((weight - from) / (to - from), 0.0, 1.0)
return t * t * (3.0 - 2.0 * t)
[/codeblock]
```
The `@GDScript` version should likely be synced with this one. Or add the details to `@GDScript`, and link to `[method @GDScript.smoothstep]` for details.

Comment (Chaosus) [reply modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: Fixed !

Comment (Chaosus) [reply modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: I changed it to the GDScript version but with example.. Seems ok ?

Comment (Chaosus) [reply modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: Oh.. wait.. I think I did it wrong

Comment (Chaosus) [reply modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml]: Hmm.. ok I changed the docs. They are more simular to each other now,  please review again

**Comment by akien-mga** [comment]
Thanks!

**Comment by hpvb** [comment]
Cherry-picked for 3.1.1.

File: doc/classes/@GDScript.xml [hunk 10]
```diff
@@ -972,6 +972,24 @@
...
+			<return type="float">
+			</return>
+			<argument index="0" name="from" type="float">
+			</argument>
+			<argument index="1" name="to" type="float">
+			</argument>
+			<argument index="2" name="weight" type="float">
+			</argument>
+			<description>
+				Returns a number smoothly interpolated between the [code]from[/code] and [code]to[/code], based on the [code]weight[/code]. Similar to [method lerp], but interpolates faster at the beginning and slower at the end.
```
Comment (kleonc) [inline doc/classes/@GDScript.xml:985]: @Chaosus, actually this implementation of **smoothstep** is similar to [inverse_lerp](https://docs.godotengine.org/en/3.2/classes/class_@gdscript.html#class-gdscript-method-inverse-lerp) and not to [lerp](https://docs.godotengine.org/en/3.2/classes/class_@gdscript.html#class-gdscript-method-lerp) so this description is misleading. Also another difference (besides interpolating faster at the beginning and slower at the end) is that **smoothstep** clamps result to range <0, 1>. So this description probably needs an update.
_(no idea if that's proper place for this note)_

**Pull request merged**